Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs

```diff
@@ -0,0 +1,151 @@
+            var currentCoin = 1;
+
+            while (currentCoin <= coin)
+            {
+                var changeArray = GenerateSingleCoinChanges(currentCoin, coins);
+                dict[currentCoin] = changeArray;
+                currentCoin++;
+            }
+
+            return dict;
+        }
+
+        public static int GetMinimalNextCoin(int coin, Dictionary<int, int[]> exchanges)
+        {
+            var nextCoin = int.MaxValue;
+            var minChange = int.MaxValue;
+
+            var coinChanges = exchanges[coin];
+
+            foreach (var change in coinChanges)
+            {
+                if (change == 0)
+                {
+                    return 0;
+                }
+
+                var currentChange = exchanges[change];
+                var min = currentChange.Min();
+
+                var minIsLesser = min < minChange;
+
+                if (minIsLesser)
+                {
+                    nextCoin = change;
+                    minChange = min;
+                }
+            }
+
+            return nextCoin;
+        }
+
+        public static int[] MakeCoinChangeDynamic(int coin, int[] coins)
+        {
+            var changesTable = GenerateChangesDictionary(coin, coins);
+            var list = new List<int>();
+
+            var currentCoin = coin;
+            var nextCoin = int.MaxValue;
+
+            while (nextCoin != 0)
+            {
+                nextCoin = GetMinimalNextCoin(currentCoin, changesTable);
+                var difference = currentCoin - nextCoin;
+                list.Add(difference);
+                currentCoin = nextCoin;
+            }
+
+            var result = list.ToArray();
+
+            return result;
+        }
+
+        private static void ValidateCoin(int coin)
+        {
+            if (coin <= 0)
+            {
+                throw new InvalidOperationException($"Coin cannot be lesser or equal to zero {nameof(coin)}.");
+            }
+        }
+
+        private static void ValidateCoinsArray(IEnumerable<int> enumerableCoins)
+        {
+            if (enumerableCoins == null)
+            {
+                throw new ArgumentNullException(nameof(enumerableCoins));
+            }
+
+            var coinsAsArray = enumerableCoins.ToArray();
+
+            if (coinsAsArray.Length == 0)
+            {
+                throw new InvalidOperationException($"Coins array cannot be empty {nameof(coinsAsArray)}.");
+            }
+
+            var coinsContainOne = coinsAsArray.Any(x => x == 1);
+
+            if (!coinsContainOne)
+            {
+                throw new InvalidOperationException($"Coins array must contain coin 1 {nameof(coinsAsArray)}.");
+            }
+
+            var containsNegative = coinsAsArray.Any(x => x <= 0);
+
+            if (containsNegative)
+            {
+                throw new InvalidOperationException(
+                    $"Coins array cannot contain negative numbers {nameof(coinsAsArray)}.");
```

Comment:
What about 0?

Reply:
Current implementation have a few requirements. 

First, it should always contain an item 1, just because of divisibility, for instance we cannot exchange 12 having coins = [3,5], because remainder 2 is lesser than minimum 3. Keeping an item 1 in the array we are able to exchange any integer.

About to keep 0, it does not make sense since that any number except 0 cannot be exchanged by 0, so it that case, for current implementation, it will be an infinit loop.